File: jbake-core/src/main/java/org/jbake/launcher/BakeWatcher.java
package org.jbake.launcher;

import org.apache.commons.configuration.CompositeConfiguration;
import org.apache.commons.vfs2.FileObject;
import org.apache.commons.vfs2.FileSystemException;
import org.apache.commons.vfs2.FileSystemManager;
import org.apache.commons.vfs2.VFS;
import org.apache.commons.vfs2.impl.DefaultFileMonitor;
import org.jbake.app.configuration.JBakeConfiguration;
import org.jbake.app.configuration.JBakeConfigurationFactory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Delegate responsible for watching the file system for changes.
 *
 * @author jmcgarr@gmail.com
 */
public class BakeWatcher {

    private Logger logger = LoggerFactory.getLogger(BakeWatcher.class);

    /**
     * Starts watching the file system for changes to trigger a bake.
     *
     * @deprecated use {@link BakeWatcher#start(JBakeConfiguration)} instead
     *
     * @param res    Commandline options
     * @param config Configuration settings
     */
    @Deprecated
    public void start(final LaunchOptions res, CompositeConfiguration config) {
        JBakeConfiguration configuration = new JBakeConfigurationFactory().createDefaultJbakeConfiguration(res.getSource(), config);
        start(configuration);
    }

    /**
     * Starts watching the file system for changes to trigger a bake.
     *
     * @param config JBakeConfiguration settings
     */
    public void start(JBakeConfiguration config) {
        try {
            FileSystemManager fsMan = VFS.getManager();
            FileObject listenPath = fsMan.resolveFile(config.getContentFolder().toURI());
            FileObject templateListenPath = fsMan.resolveFile(config.getTemplateFolder().toURI());
            FileObject assetPath = fsMan.resolveFile(config.getAssetFolder().toURI());

            logger.info("Watching for (content, template, asset) changes in [{}]", config.getSourceFolder().getPath());
            DefaultFileMonitor monitor = new DefaultFileMonitor(new CustomFSChangeListener(config));
            monitor.setRecursive(true);
            monitor.addFile(listenPath);
            monitor.addFile(templateListenPath);
            monitor.addFile(assetPath);
            monitor.start();
        } catch (FileSystemException e) {
            logger.error("Problems watching filesystem changes", e);
        }
    }
}

File: jbake-core/src/main/java/org/jbake/launcher/BakeWatcher.kt
package org.jbake.launcher

import org.apache.commons.vfs2.FileSystemException
import org.apache.commons.vfs2.VFS
import org.apache.commons.vfs2.impl.DefaultFileMonitor
import org.jbake.app.configuration.JBakeConfiguration
import org.slf4j.Logger
import org.jbake.util.Logging.logger

/**
 * Delegate responsible for watching the file system for changes.
 */
class BakeWatcher {


    /**
     * Starts watching the file system for changes to trigger a bake.
     */
    fun start(config: JBakeConfiguration) {
        try {
            val fsMan = VFS.getManager()
            val listenPath = fsMan.resolveFile(config.contentDir.toURI())
            val templateListenPath = fsMan.resolveFile(config.templateDir.toURI())
            val assetPath = fsMan.resolveFile(config.assetDir.toURI())
            val dataPath = fsMan.resolveFile(config.dataDir.toURI())

            log.info("Watching for (content, data, template, asset) changes in [{}]", config.sourceDir!!.path)
            val monitor = DefaultFileMonitor(CustomFileSystemChangeListener(config))
            monitor.isRecursive = true
            monitor.addFile(listenPath)
            monitor.addFile(templateListenPath)
            monitor.addFile(assetPath)
            monitor.addFile(dataPath)
            monitor.start()
        }
        catch (e: FileSystemException) { log.error("Problems watching filesystem changes", e) }
    }

    private val log: Logger by logger()
}
